drivers and samples: Infineon CAT1 counter driver

pavloinfineon opened the pull request:
Added initial version of Infineon CAT1 counter driver
Added external trigger pin that is able to run counter
Added Pulse Counter sample

ifyall commented on samples/boards/infineon/pulse_counter/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+sample:
+  name: Pulse Counter Driver Test
```
```suggestion
  name: Pulse Counter Driver Example
```

ifyall commented on drivers/counter/CMakeLists.txt:
```diff
@@ -37,8 +37,8 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_XLNX_AXI_TIMER      counter_xlnx_axi
 zephyr_library_sources_ifdef(CONFIG_COUNTER_TMR_ESP32           counter_esp32_tmr.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_RTC_ESP32           counter_esp32_rtc.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MICROCHIP_MCP7940N  rtc_mcp7940n.c)
-zephyr_library_sources_ifdef(CONFIG_COUNTER_ANDES_ATCPIT100     counter_andes_atcpit100.c)
```
Should not remove a non-Infineon driver.

fabiobaltieri commented on drivers/counter/Kconfig.ifx_cat1:
```diff
@@ -0,0 +1,13 @@
+# Infineon CAT1 counter driver
+
+# Copyright (c) 2022 Cypress Semiconductor Corporation (an Infineon company) or
```
2023?

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,488 @@
+};
+
+/* Default Counter configuration structure */
+const cy_stc_tcpwm_counter_config_t _cyhal_timer_default_config = {
```
Drop the `_` prefix, we don't use it in Zephyr. Also make it `static`.

pavloinfineon replied:
Updated

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,488 @@
+	.countInput = CY_TCPWM_INPUT_1,
+};
+
+static int32_t _get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
```
Drop the `_` prefixes.

pavloinfineon replied:
Updated

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,488 @@
+
+static int32_t _get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
+{
+	int32_t ret = -1;
```
How about using a more meaningful error code?

pavloinfineon replied:
Changed to EINVAL

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,488 @@
+
+	cyhal_timer_enable_event(&data->counter_obj, CYHAL_TIMER_IRQ_CAPTURE_COMPARE,
+				 config->irq_priority, true);
+	NVIC_SetPendingIRQ((IRQn_Type)config->irqn);
```
Would it make more sense to have `irqn` of the right type in the struct and only cast it once on the initial assignment?

pavloinfineon replied:
Changed `config->irqn` to IRQn_TYPE

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,488 @@
+	cy_rslt_t rslt;
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+	bool ticks_gt_period = false;
```
No need to initialize this one.

pavloinfineon replied:
Fixed

fabiobaltieri commented on samples/boards/infineon/pulse_counter/src/main.c:
```diff
@@ -0,0 +1,124 @@
+
+void main(void)
+{
+	int ret;
+
+	printk("Pulse Counter Sample\n\n");
+
+	const struct device *p_dev_counter1 = COUNTER_1;
+
+	int err;
```
this looks a bit odd, do the declaration first and printk after?

pavloinfineon replied:
Fixed order and moved printk

fabiobaltieri commented on samples/boards/infineon/pulse_counter/src/main.c:
```diff
@@ -0,0 +1,124 @@
+		return;
+	}
+
+	if (!device_is_ready(pulse.port)) {
```
You can use `gpio_is_ready_dt` here

pavloinfineon replied:
Changed

fabiobaltieri commented on samples/boards/infineon/pulse_counter/src/main.c:
```diff
@@ -0,0 +1,124 @@
+	printk("!! Alarm !!\n");
+	printk("!! Count = %u\n", ticks);
+
+	struct counter_alarm_cfg *p_alarm = (struct counter_alarm_cfg *)user_data;
```
Same as in main, keep these at the top, not sure it's explicitly in the coding style but that's how most of the code is written.

pavloinfineon replied:
Fixed. Also changed in a couple other functions

fabiobaltieri commented on samples/boards/infineon/pulse_counter/src/main.c:
```diff
@@ -0,0 +1,124 @@
+#define PULSE_PIN  1
+#define PULSE_NODE DEVICE_DT_GET(DT_NODELABEL(gpio_prt9))
+
+struct counter_top_cfg counter_top;
+int counter_top_userdata;
```
static?

pavloinfineon replied:
Changed

pavloinfineon commented:
twister build 2 CI is failing compiling `drivers/counter/counter_ll_stm32_timer.c` which is unrelated to changes done in the recent commits. Double checked that kconfig or drivers.counter.basic_api changes look to be as expected so unsure what is causing these errors.
I was also unable to reproduce these compile errors locally.

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,537 @@
+	for (i = 0u; i < _CYHAL_TCPWM_INSTANCES; i++) {
+		TCPWM_Type *base = _CYHAL_TCPWM_DATA[i].base;
+
+		if (((uint32_t)reg_addr > (uint32_t)base) &&
+		    ((uint32_t)reg_addr < (uint32_t)base + sizeof(TCPWM_Type))) {
```
Hey I think that a cleaner way to do this would be to use `uintptr_t` and `POINTER_TO_UINT`, you could initialize `base` as:
```c
uintptr_t base = POINTER_TO_UINT(_CYHAL_TCPWM_DATA[i].base)
```

add a temporary variable for `reg_addr` and I guess `cnt`, and then you would not have to carry castings around the whole function.

pavloinfineon replied:
Simplified the casts around the function

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,537 @@
+
+	if ((absolute && (val < curr)) || (diff > max_rel_val)) {
+		if (absolute) {
+			err = -ETIME;
```
The exit path of this function is a bit hard to follow, how about:
- change the function exit on 445 to just `return 0`
- move this after the `irq_on_late` check and do a `return -ETIME` directly (the function behavior is the same)
- drop `err`, you won't need it anymore

I think that would make it more obvious to understand what's going on.

pavloinfineon replied:
Changed in most recent push

fabiobaltieri commented:
> Twister build 2 CI is failing compiling `drivers/counter/counter_ll_stm32_timer.c` which is unrelated to changes done in the recent commits.

Yeah it was broken upstream, should be fixed now, sorry about it, in this situation you could check on #ci in Discord or check recent PRs to see if someone is working on it or wait it out and force push again to trigger another CI run. (or send a fix :-))

gmarull reviewed (dismissed):
Board specific samples should rarely exist when using driver APIs (the purpose of them is to be able to write portable applications, actually).

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,539 @@
+			hw_resource->channel_num = ((reg_addr_ptr - cnt) / sizeof(TCPWM_CNT_Type));
+
+			if (hw_resource->channel_num < _CYHAL_TCPWM_DATA[i].num_channels) {
+				ret = 0;
+			}
+			break;
```
The error path here is quite hard to follow, if I understand the intent could you change it to:
- `return 0` on like 87
- `return -EINVAL` on line 89 and 93

That should be easier to understand. Even better flip the condition on line 86 and make that the error path.

pavloinfineon replied:
Flipped the conditions to make it easier to understand

pavloinfineon commented:
> Board specific samples should rarely exist when using driver APIs (the purpose of them is to be able to write portable applications, actually).

Removed our board specific sample

fabiobaltieri commented on drivers/counter/counter_ifx_cat1.c:
```diff
@@ -0,0 +1,537 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ifx_cat1_counter, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Device config structure */
```
nit: I would drop this and the ones in line 34 and 46, the struct names are meaningful enough, no need to repeat those in comments, up to you though

pavloinfineon replied:
You were correct, comments did seem redundant. Removed them

pavloinfineon committed:
drivers: counter: Add Infineon CAT1 counter driver

Add initial version of Infineon CAT1 counter driver
Add initial version of binding file for Infineon
Add counters to psco6 dtsi
Add external trigger pin that runs counter

Signed-off-by: Pavlo Havrylyuk <pavlo.havrylyuk@infineon.com>
